File: lib/python/vplaned/vplaned.py
# Copyright (c) 2017-2018, AT&T Intellectual Property. All rights reserved.
# Copyright (c) 2015 Brocade Communications Systems, Inc.
# All Rights Reserved.
#
# SPDX-License-Identifier: LGPL-2.1-only
"""Python3 vplane-controller API.

This module exports 2 classes, Controller and Dataplane.
The Controller can be used to store config command to the dataplane(s), or to
generate a Dataplane object for each connected dataplane(s) (local/remote).
The Dataplane object will have all the information returned by the Controller
as attributes. EG: dataplane.local, dataplane.control, etc.
The "interfaces" attribute is a dictionary keyed on interface name associated
with a corresponding Interface object, which similarly holds state in
attributes like index, state, name, etc.
Generate is intended in the Python sense, so the Dataplane class should not be
instantiated manually. Instead the Controller.get_dataplanes() generator should
be used. Examples:

from vplaned import Controller

with Controller() as controller:
    for dp in controller.get_dataplanes():
        for name, intf in dp.interfaces.items():
            intf.state
        with dp:
            dp.json_command("netflow show")

with Controller(store_endpoint="tcp://1.2.3.4:5678") as controller:
    controller.store("interface dataplane netflow dp1s12",
                     "netflow disable dp1s12")

Note that the store function now supports two messaging modes: the old space
delimited text command (as shown above). And the 'new' messaging mode using
protocol buffers. The protocol buffers use of the store() function requires an
additional argument: cmd_name.

In case of errors with the controller or dataplane runtime commands,
ControllerException or DataplaneException will be raised respectively.
In case of connectivity errors, ZMQError will be raised.

Note that thanks to Python's ContextManager pattern, no cleanup/disconnect is
required. The "with" statement will take care of everything automagically.
"""
import sys
import os
import re
import zmq

class InterfaceException(Exception):
    pass


class DataplaneException(Exception):
    pass


class ControllerException(Exception):
    pass


class Interface:

    """Interface object. This will be automatically generated and added to a
    Dataplane object's "interfaces" attribute dictionary, keyed by the name."""

    def __init__(self, json):
        for key, value in json.items():
            setattr(self, key, value)

        self.dp_id = Interface.get_dp_id(self.name)
        if self.dp_id is None:
            raise InterfaceException("Invalid name: {}".format(self.name))

    @classmethod
    def get_dp_id(cls, interface_name):
        """Parse an interface name and return the dataplane ID as an integer"""
        m = re.match("^[a-z]+(\d+)", interface_name)
        if m is None:
            return None

        return int(m.group(1))


class Dataplane:

    """Dataplane object. It is strongly recommended to avoid creating this
    object manually. Instead, use the Controller.get_dataplanes() generator.
    All the information that vplane-controller has about a dataplane will be
    stored in this object as attributes. EG: dp.local, dp.control, etc.
    The "interfaces" attribute will be a dictionary with interface names as key
    and Interface objects as values.

    Implements ContextManager pattern, so use through "with" statement."""

    def __init__(self, ctx, json):
        self._ctx = ctx
        self._socket = None
        self.interfaces = {}
        for key, value in json.items():
            # The interfaces json object is treated differently by the
            # controller, it returns as an array of flat json objects so we
            # need this bit of magic string matching to make it work
            if key == "interfaces":
                for intf in value:
                    self.interfaces[intf["name"]] = Interface(intf)
            else:
                setattr(self, key, value)

    def __enter__(self):
        self._socket = self._ctx.socket(zmq.REQ)
        # Dataplane/Controller in 2.0.0 do not return "control"
        if hasattr(self, "control"):
            self._socket.connect(self.control)
        else:
            self._socket.connect("ipc:///var/run/vplane.socket")
        return self

    def __exit__(self, *exc):
        self._socket.close()

    def string_command(self, string):
        """send a command and return the dataplane response as a string"""
        self._socket.send_string(string)
        rc = self._socket.recv_string()
        if rc != "OK":
            raise DataplaneException("Command {} returned {}".format(string,
                                                                     rc))

        return self._socket.recv_string()

    def json_command(self, string):
        """send a command and return the dataplane response as a json object"""
        self._socket.send_string(string)
        rc = self._socket.recv_string()
        if rc != "OK":
            raise DataplaneException("Command {} returned {}".format(string,
                                                                     rc))

        return self._socket.recv_json()


class Controller:

    """Controller object. Can be used to generate Dataplane objects or to store
    config.
    Implements ContextManager pattern, so use through "with" statement.
    """

    def __init__(self, store_endpoint="ipc:///var/run/vyatta/vplaned.socket",
                 cfg_endpoint="ipc:///var/run/vyatta/vplaned-config.socket"):
        self._ctx = None
        self._store_socket = None
        self._cfg_socket = None
        self._store_endpoint = store_endpoint
        self._cfg_endpoint = cfg_endpoint

    def __enter__(self):
        self._ctx = zmq.Context.instance()
        # ZMQ sockopts set in the context will be used by all sockets
        self._ctx.IPV6 = 1
        self._ctx.RCVTIMEO = 10000
        # All messages are ACK'ed so no need to wait when the socket is closed
        self._ctx.LINGER = 0
        self._store_socket = self._ctx.socket(zmq.REQ)
        self._store_socket.connect(self._store_endpoint)
        self._cfg_socket = self._ctx.socket(zmq.REQ)
        self._cfg_socket.connect(self._cfg_endpoint)
        return self

    def __exit__(self, *exc):
        self._store_socket.close()
        self._cfg_socket.close()

    def get_dataplanes(self):
        """Generator for dataplanes. Will fetch info from the controller about
        all dataplanes, and create an object each and yield it.
        """
        self._cfg_socket.send_string("GETVPCONFIG")
        rc = self._cfg_socket.recv_string()
        if rc != "OK":
            raise ControllerException("GETVPCONFIG returned {}".format(rc))

        json = self._cfg_socket.recv_json()
        if json is None:
            raise ControllerException("GETVPCONFIG returned empty response")

        for dp in json["dataplanes"]:
            yield Dataplane(self._ctx, dp)

    def store(self, path, cmd, interface="ALL",
              action=os.getenv("COMMIT_ACTION"),
              cmd_name=None):
        """Send command to dataplane(s) and store it, associated with the path.
        By default it will apply to all interfaces, "interface" parameter to
        override.
        By default it will use the action defined by the COMMIT_ACTION env
        variable, which is set when in commit mode, "action" parameter to
        override.
        """
        if action is None:
            raise ControllerException(
                "COMMIT_ACTION not found. Not in commit mode?")

        # Create a dictionary, and use a reference to work on it recursively.
        # The dataplane expect the path to be passed as nested JSON objects,
        # with the actual command as a value in the bottom object.
        msg = {}
        temp = msg
        for item in path.split(" "):
            temp = temp.setdefault(item, {})
        temp["__" + action + "__"] = cmd
        temp['__INTERFACE__'] = interface

        # cmd_name if defined is used to send a protocol buffers form
        # of the command.
        if (cmd_name is not None):
            import base64
            import google.protobuf
            import vyatta.proto.DataplaneEnvelope_pb2
            import vyatta.proto.VPlanedEnvelope_pb2

            # Build up the Dataplane Envelope here
            de = vyatta.proto.DataplaneEnvelope_pb2.DataplaneEnvelope()
            de.type = cmd_name
            de.msg = cmd.SerializeToString()
            
            # Build up the VPlaned Envelope here
            ve = vyatta.proto.VPlanedEnvelope_pb2.VPlanedEnvelope()
            ve.key = path
            ve.interface = interface
            
            if (action == "SET"):
                ve.action = vyatta.proto.VPlanedEnvelope_pb2.VPlanedEnvelope.SET
            else:
                ve.action = vyatta.proto.VPlanedEnvelope_pb2.VPlanedEnvelope.DELETE
                
            ve.msg = de.SerializeToString()

            # Convert to base64
            temp["__PROTOBUF__"] = base64.b64encode(ve.SerializeToString())
        
        self._store_socket.send_json(msg)
        rc = self._store_socket.recv_string()
        if rc != "OK":
            raise ControllerException("Config {} returned {}".format(msg, rc))

    def config(self, cmd):
        self._cfg_socket.send_multipart(cmd)
        rc = self._cfg_socket.recv_string()
        if rc != "OK":
            raise ControllerException("Config cmd returned {}".format(rc))

File: lib/go/vplaned/vplaned.go
// Copyright (c) 2020 , AT&T Intellectual Property. All rights reserved.
// SPDX-License-Identifier: MPL-2.0

package vplaned

import (
	"encoding/base64"
	"encoding/json"
	"os"
	"runtime"
	"strings"
	"sync"

	"github.com/danos/vyatta-controller/protobuf/go/VPlanedEnvelope"
	"github.com/danos/vyatta-dataplane/protobuf/go/DataplaneEnvelope"
	"github.com/golang/protobuf/proto"
	zmq "github.com/zeromq/goczmq"
)

const (
	storeEndpoint  = "ipc:///var/run/vyatta/vplaned.socket"
	configEndpoint = "ipc:///var/run/vyatta/vplaned-config.socket"
)

type Error string

func (e Error) Error() string {
	return string(e)
}

const (
	ErrNoControllerResponse = Error("No response from controller")
	ErrConfigStoreFailed = Error("Config store failed")
	ErrNoCommitAction = Error("COMMIT_ACTION not found")
)

type Conn struct {
	mu sync.RWMutex

	storeEndpoint, configEndpoint string
	configSocket, storeSocket     *zmq.Sock
}

type dialOpt func(*Conn)

func ConfigEndpoint(address string) dialOpt {
	return func(c *Conn) {
		c.configEndpoint = address
	}
}

func StoreEndpoint(address string) dialOpt {
	return func(c *Conn) {
		c.storeEndpoint = address
	}
}

func Dial(opts ...dialOpt) (*Conn, error) {
	out := Conn{
		storeEndpoint:  storeEndpoint,
		configEndpoint: configEndpoint,
	}
	for _, opt := range opts {
		opt(&out)
	}
	storeSock, err := zmq.NewReq(
		out.storeEndpoint,
		zmq.SockSetLinger(0),
		zmq.SockSetRcvtimeo(10000),
		zmq.SockSetIpv6(1),
	)
	if err != nil {
		return nil, err
	}
	confSock, err := zmq.NewReq(
		out.configEndpoint,
		zmq.SockSetLinger(0),
		zmq.SockSetRcvtimeo(10000),
		zmq.SockSetIpv6(1),
	)
	if err != nil {
		return nil, err
	}

	out.storeSocket = storeSock
	out.configSocket = confSock
	runtime.SetFinalizer(&out, func(c *Conn) {
		c.Close()
	})
	return &out, nil
}

func (c *Conn) Close() error {
	c.mu.Lock()
	defer c.mu.Unlock()

	if c.storeSocket == nil || c.configSocket == nil {
		return nil
	}
	err := c.storeSocket.Disconnect(c.storeEndpoint)
	if err != nil {
		return err
	}
	err = c.configSocket.Disconnect(c.configEndpoint)
	if err != nil {
		return err
	}
	c.storeSocket.Destroy()
	c.configSocket.Destroy()
	c.storeSocket = nil
	c.configSocket = nil
	return nil
}

func (c *Conn) Store(
	path, msgType string,
	msg proto.Message,
	opts ...storeOpt,
) error {
	c.mu.RLock()
	defer c.mu.RUnlock()

	storeOpts := storeOptions{}
	for _, opt := range opts {
		opt(&storeOpts)
	}
	if storeOpts.action == "" {
		storeOpts.action = os.Getenv("COMMIT_ACTION")
	}

	if storeOpts.action == "" {
		return ErrNoCommitAction
	}

	msgData, err := proto.Marshal(msg)
	if err != nil {
		return err
	}

	de := DataplaneEnvelope.DataplaneEnvelope{
		Type: proto.String(msgType),
		Msg:  msgData,
	}

	deData, err := proto.Marshal(&de)
	if err != nil {
		return err
	}

	vpe := VPlanedEnvelope.VPlanedEnvelope{
		Key: proto.String(path),
		Msg: deData,
	}
	if storeOpts.iface != "" {
		vpe.Interface = proto.String(storeOpts.iface)
	}
	if storeOpts.action == "SET" {
		vpe.Action = VPlanedEnvelope.VPlanedEnvelope_SET.Enum()
	} else {
		vpe.Action = VPlanedEnvelope.VPlanedEnvelope_DELETE.Enum()
	}
	vpeData, err := proto.Marshal(&vpe)
	if err != nil {
		return err
	}

	cmd := "protobuf " + base64.StdEncoding.EncodeToString(vpeData)

	jobj := make(map[string]interface{})
	tmp := jobj
	for _, elem := range strings.Split(path, " ") {
		obj := make(map[string]interface{})
		tmp[elem] = obj
		tmp = obj
	}
	tmp["__"+storeOpts.action+"__"] = cmd
	if storeOpts.iface != "" {
		tmp["__INTERFACE__"] = storeOpts.iface
	}
	tmp["__PROTOBUF__"] = true

	encodedMsg, err := json.Marshal(jobj)
	if err != nil {
		return err
	}

	err = c.storeSocket.SendMessage([][]byte{encodedMsg})
	if err != nil {
		return err
	}
	zmsg, err := c.storeSocket.RecvMessage()
	if err != nil {
		return err
	}
	if len(zmsg) == 0 {
		return ErrNoControllerResponse
	}
	if string(zmsg[0]) != "OK" {
		return ErrConfigStoreFailed
	}
	return nil
}

type storeOpt func(*storeOptions)
type storeOptions struct {
	action, iface string
}

func Action(action string) storeOpt {
	return func(o *storeOptions) {
		o.action = action
	}
}

func Interface(iface string) storeOpt {
	return func(o *storeOptions) {
		o.iface = iface
	}
}
